Wrappers/Python/ccpi/viewer/utils/io.py

```diff
@@ -42,3 +66,765 @@ def SaveRenderToPNG(render_window, filename):
+
+    def AddChildDataset(self, child_dataset, attributes=None):
+        if not isinstance(child_dataset, vtk.vtkImageData):
+            raise Exception("child_dataset must be vtk.vtkImageData")
+        # check type is vtkImageData
+        self._ValidateChildDatasetAttributes(child_dataset, attributes)
+        self._ChildDatasets.append(child_dataset)
+        self._ChildDatasetsAttributes.append(attributes)
+
+    def _ValidateChildDatasetAttributes(self, child_dataset, attributes):
+        if not isinstance(attributes, dict) and not (attributes is None):
+            raise Exception("'attributes' must be a dictionary, or unset (i.e. None)")
+
+    def SetChunking(self, chunking):
+        '''
+        Parameters
+        ----------
+        chunking: bool, default: True in case of HDF5 file,
+                  otherwise False
+            Whether to write to the file in chunks.
+        '''
+        self._Chunking = chunking
+
+    def SetChunkShape(self, chunk_shape):
+        '''
+        Parameters
+        ----------
+        chunk_shape: tuple, default: one slice on the z axis
+            If SetChunking() has been set to True, this will be
+            the size of a chunk written to the file.
+        '''
+        self._ChunkShape = chunk_shape
+
+    def SetHDF5Compression(self, compression):
+        '''
+        Parameters
+        ----------
+        compression: list, default None
+            This is a list of compression settings when writing to HDF5
+            First element: str
+                The type of hdf5 compression to use.
+                Must be one of: None, 'gzip', or 'lzf'
+            Second element:
+                Options for the compression filter. For details, see:
+                https://docs.h5py.org/en/stable/high/dataset.html?highlight=compression_opts#dataset-compression
+            Third element: bool
+                Whether to shuffle the data i.e. rearrange the bytes in
+                a chunk, which may improve the compression ratio
+        '''
+        self._HDF5Compression = compression
+
+    def Write(self):
+        # check file ext
+        writer = self._GetWriter()
+        writer.Write()
+
+    def _GetWriter(self):
+        file_name = os.path.splitext(self._FileName)[0]
+
+        if self._FileFormat in ['nxs', 'h5', 'hdf5', '']:
+            self._FileName = file_name + '.hdf5'
+            writer = self._GetHDF5Writer()
+
+        elif self._FileFormat in ['mha']:
+            self._FileName = file_name + '.mha'
+            writer = self._GetMetaImageWriter()
+
+        else:
+            raise Exception("File format is not supported. Supported types include hdf5/nexus.")
+
+        writer.SetFileName(self._FileName)
+        return writer
+
+    def _GetHDF5Writer(self):
+        writer = cilviewerHDF5Writer()
+        writer.SetOriginalDataset(None, self._OriginalDatasetAttributes)
+        writer.SetChunking(self._Chunking)
+        writer.SetChunkShape(self._ChunkShape)
```

Comment:
What is the chunk shape set to? It seems we ask for chunking but not set the shape.

Reply:
There is a method SetChunkShape

Reply:
(in the ImageWriter) I mean. By default this is None. If the chunk shape is None in the cilviewerHDF5Writer then it uses one slice on the z axis as the chunk shape by default

Reply:
(if chunking has been set to True)

Reply:
I have also made a new class called ImageWriterInterface and moved methods used by both the ImageWriter and cilviewerHDF5Writer into there to make things tidier, more logical, and less confusing.

Reply:
That's right